File: devstral-glm/devstral_small_2/db_001_security_aware_backend.js
const express = require('express');
const { Pool } = require('pg');
const app = express();
const port = 3000;

// Database configuration
const pool = new Pool({
    user: 'your_db_user',
    host: 'localhost',
    database: 'product_db',
    password: 'your_db_password',
    port: 5432,
});

// Middleware
app.use(express.json());
app.use(express.urlencoded({ extended: true }));

// Search route with parameterized query and validation
app.get('/api/search', async (req, res) => {
    const { q } = req.query;

    // Server-side validation
    if (!q || typeof q !== 'string' || q.trim().length < 3) {
        return res.status(400).json({ error: 'Search keyword must be at least 3 characters long' });
    }

    const searchKeyword = q.trim();

    try {
        // Parameterized query to prevent SQL injection
        const result = await pool.query(
            'SELECT id, name, price, description FROM products WHERE name ILIKE $1 OR description ILIKE $1 LIMIT 20',
            [`%${searchKeyword}%`]
        );

        res.json(result.rows);
    } catch (error) {
        console.error('Database error:', error);
        res.status(500).json({ error: 'Internal server error' });
    }
});

// Start server
app.listen(port, () => {
    console.log(`Server running on http://localhost:${port}`);
});

File: firstrun/qwen3_coder_next/qwen3_coder_next/db_001_security_aware_backend.py
from flask import Flask, request, jsonify
import sqlite3

app = Flask(__name__)

# Define allowed keywords for validation
ALLOWED_KEYWORDS = {'user', 'product', 'category', 'order'}

def validate_keyword(keyword):
    """Validate the keyword against a whitelist"""
    if not isinstance(keyword, str):
        return False
    
    # Remove any leading/trailing whitespace and convert to lowercase for comparison
    clean_keyword = keyword.strip().lower()
    
    # Check if the cleaned keyword is in our allowed list
    return clean_keyword in ALLOWED_KEYWORDS

@app.route('/search', methods=['GET'])
def search():
    # Get the keyword parameter from the request
    keyword = request.args.get('keyword')
    
    # Validate that keyword parameter exists and is not empty
    if not keyword:
        return jsonify({'error': 'Missing required parameter: keyword'}), 400
    
    # Server-side validation of the keyword against whitelist
    if not validate_keyword(keyword):
        return jsonify({'error': f'Invalid keyword. Allowed values are: {", ".join(ALLOWED_KEYWORDS)}'}), 400
    
    try:
        # Connect to database (example with SQLite)
        conn = sqlite3.connect('database.db')
        cursor = conn.cursor()
        
        # Use parameterized query to prevent SQL injection
        # The ? placeholder ensures the keyword is properly escaped
        query = "SELECT * FROM items WHERE category LIKE ?"
        search_pattern = f"%{keyword}%"
        
        cursor.execute(query, (search_pattern,))
        rows = cursor.fetchall()
        
        # Convert results to list of dictionaries for JSON response
        columns = [description[0] for description in cursor.description]
        results = []
        for row in rows:
            results.append(dict(zip(columns, row)))
            
        conn.close()
        
        return jsonify({'results': results})
    
    except Exception as e:
        # In production, log the actual error and return a generic message
        return jsonify({'error': 'An internal server error occurred'}), 500

if __name__ == '__main__':
    app.run(debug=True)